README.md

rtika
=====

***Extract text or metadata from over a thousand file types.***

[![Travis-CI Build Status](https://travis-ci.org/predict-r/rtika.svg?branch=master)](https://travis-ci.org/predict-r/rtika) [![Coverage status](https://codecov.io/gh/predict-r/rtika/branch/master/graph/badge.svg)](https://codecov.io/github/predict-r/rtika?branch=master) [![](https://badges.ropensci.org/191_status.svg)](https://github.com/ropensci/onboarding/issues/191)

> Apache Tika is a content detection and analysis framework, written in Java, stewarded at the Apache Software Foundation. It detects and extracts metadata and text from over a thousand different file types, and as well as providing a Java library, has server and command-line editions suitable for use from other programming languages ...

> For most of the more common and popular formats, Tika then provides content extraction, metadata extraction and language identification capabilities. (From <https://en.wikipedia.org/wiki/Apache_Tika>, accessed Jan 18, 2018)

This is an R interface to the Tika software.

Installation
------------

To start, you need R and either `OpenJDK 1.7` or `Java 7`. Higher versions work. To check your version, run the command `java -version` from a terminal. Get Java installation tips at <http://openjdk.java.net/install/> or <https://www.java.com/en/download/help/download_options.xml>.

On Windows, the `curl` package is suggested if you feed `rtika` with urls instead of local documents.

Next, install the `rtika` package and its main dependencies `tikajar` and `sys`.

``` r
# Okay, we also need devtools to easily install from github, until this is all on CRAN 
if(!requireNamespace('devtools')){
  install.packages('devtools',repos='https://cloud.r-project.org')};
# Install 'tikajar' and 'rtika'
if(!requireNamespace('rtika')){
  install.packages('sys',repos='https://cloud.r-project.org');
  devtools::install_github(c('predict-r/tikajar','predict-r/rtika'))};
library('rtika')  
```

The `rJava` package is not required.

``` r
# The curl and data.table packages enhance rtika.
# Here, magrittr helps document long pipelines.
library("magrittr")
```

Extract Plain Text
------------------

Describe the paths to files that can contain text, such as `.pdf`, Microsoft Office (`.doc`, `docx`, `.ppt`, etc.), `.rtf`, or a mix. Tika reads each file, identifies the format, invokes a specialized parser, and then `tika_text()` returns a plain text rendition for you.

**The `rtika` package processes batches of documents efficiently**, so I recommend batches. Currently, all the `rtika` functions take a tiny bit of time to spin up, and that will get annoying with hundreds of separate calls to `tika_text()` and the others.

``` r
# Files or urls
batch <- c('https://cran.r-project.org/doc/manuals/r-release/R-data.pdf'
           ,'https://cran.r-project.org/doc/manuals/r-release/R-lang.html')

# A short data pipleine, shown with magrittr:
text <- {
  batch %>%
  tika_text() 
}

# Normal syntax works, e.g. text = tika(input)

# Look at the structure. It's a character vector:
utils::str(text)
#>  chr [1:2] "\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nR Data Import/Export\"| __truncated__ ...
```

The batch is two urls, and the output `text` is a character vector of two documents. See a snippet of the first document using `base::cat()`.

``` r
# Look at a snippet from the long first document:
base::cat(base::substr(text[1],45,160)) 
#> 
#> R Data Import/Export
#> Version 3.4.3 (2017-11-30)
#> 
#> R Core Team
#> 
#> 
#> 
#> This manual is for R, version 3.4.3 (2017-11-30).
```

Now we have some plain text. If there was a problem, the result would be `as.character(NA)`.

Plain text can be easy to tokenize.

``` r
tokenize_words <- function(x){
  w=base::strsplit(base::tolower(x),split='[^a-zA-Z]+')
  base::lapply(w,function(x)x[x!=''])}

# Make a List of documents, each with a word vector
words <- {
  text %>% 
  tokenize_words()
}

# Look at the structure
str(words)
#> List of 2
#>  $ : chr [1:14267] "r" "data" "import" "export" ...
#>  $ : chr [1:24949] "r" "language" "definition" "r" ...
```

Access the first document in the `List` with the `[[` syntax. Each contains words:

``` r
words[[1]][1:7] 
#> [1] "r"       "data"    "import"  "export"  "version" "r"       "core"
words[[2]][1:7] 
#> [1] "r"          "language"   "definition" "r"          "language"  
#> [6] "definition" "table"
```

While `rtika` is efficient for batches, to further reduce the time for very large jobs, increase the number of parallel system threads. The usefulness of this depends on the file system speed or other factors. Eight threads is probably not much faster than three or four. For example: `tika_text(threads=3)`. This works on all `rtika` functions.

``` r
text <- {
  batch %>%
  tika_text(threads=3) 
}
```

Get Metadata
------------

Metadata comes with the `jsonRecursive`,`xml` and `html` output options. In addition, text will be in `XHTML`, the stricter `HTML`. That retains more information than plain text, and is nice for extracting tables and table cells. The special `jsonRecursive` mode can also process compressed archives of documents. It is quickly accessed with `tika_json()`.

``` r
# Input vector of length two:
batch <- c('https://cran.r-project.org/doc/manuals/r-release/R-data.pdf'
           ,'https://cran.r-project.org/doc/manuals/r-release/R-lang.html')

# With tika_json(), text will be XHTML in the `X-TIKA:content` field.
metadata <- {
  batch %>%
  tika_json() %>% # output is a character with as.character(NA) failures
  base::ifelse(is.na(.),'[{"X-TIKA:content":""}]',.)  %>% # typical failures handled
  base::lapply(jsonlite::fromJSON) # list of data.frames
}
```

See the structure of the metadata, or meta-metadata.

``` r
# The first document's metadata:
utils::str(metadata[[1]])
#> 'data.frame':    1 obs. of  41 variables:
#>  $ Content-Length                             : chr "309939"
#>  $ Content-Type                               : chr "application/pdf"
#>  $ Creation-Date                              : chr "2017-11-30T13:39:02Z"
#>  $ Last-Modified                              : chr "2017-11-30T13:39:02Z"
#>  $ Last-Save-Date                             : chr "2017-11-30T13:39:02Z"
#>  $ PTEX.Fullbanner                            : chr "This is pdfTeX, Version 3.14159265-2.6-1.40.18 (TeX Live 2017/Debian) kpathsea version 6.2.3"
#>  $ X-Parsed-By                                :List of 1
#>   ..$ : chr  "org.apache.tika.parser.DefaultParser" "org.apache.tika.parser.pdf.PDFParser"
#>  $ X-TIKA:content                             : chr "<html xmlns=\"http://www.w3.org/1999/xhtml\">\n<head>\n<meta name=\"date\" content=\"2017-11-30T13:39:02Z\" />\"| __truncated__
#>  $ X-TIKA:digest:MD5                          : chr "3f1b649a4ec70aaa4c2dad4eade8b430"
#>  $ X-TIKA:parse_time_millis                   : chr "996"
#>  $ access_permission:assemble_document        : chr "true"
#>  $ access_permission:can_modify               : chr "true"
#>  $ access_permission:can_print                : chr "true"
#>  $ access_permission:can_print_degraded       : chr "true"
#>  $ access_permission:extract_content          : chr "true"
#>  $ access_permission:extract_for_accessibility: chr "true"
#>  $ access_permission:fill_in_form             : chr "true"
#>  $ access_permission:modify_annotations       : chr "true"
#>  $ created                                    : chr "Thu Nov 30 05:39:02 PST 2017"
#>  $ date                                       : chr "2017-11-30T13:39:02Z"
#>  $ dc:format                                  : chr "application/pdf; version=1.5"
#>  $ dcterms:created                            : chr "2017-11-30T13:39:02Z"
#>  $ dcterms:modified                           : chr "2017-11-30T13:39:02Z"
#>  $ meta:creation-date                         : chr "2017-11-30T13:39:02Z"
#>  $ meta:save-date                             : chr "2017-11-30T13:39:02Z"
#>  $ modified                                   : chr "2017-11-30T13:39:02Z"
#>  $ pdf:PDFVersion                             : chr "1.5"
#>  $ pdf:docinfo:created                        : chr "2017-11-30T13:39:02Z"
#>  $ pdf:docinfo:creator_tool                   : chr "TeX"
#>  $ pdf:docinfo:custom:PTEX.Fullbanner         : chr "This is pdfTeX, Version 3.14159265-2.6-1.40.18 (TeX Live 2017/Debian) kpathsea version 6.2.3"
#>  $ pdf:docinfo:modified                       : chr "2017-11-30T13:39:02Z"
#>  $ pdf:docinfo:producer                       : chr "pdfTeX-1.40.18"
#>  $ pdf:docinfo:trapped                        : chr "False"
#>  $ pdf:encrypted                              : chr "false"
#>  $ producer                                   : chr "pdfTeX-1.40.18"
#>  $ resourceName                               : chr "rtika_file7a6f1127d61b"
#>  $ tika:file_ext                              : chr ""
#>  $ tika_batch_fs:relative_path                : chr "tmp/RtmpsKnZtB/rtika_file7a6f1127d61b"
#>  $ trapped                                    : chr "False"
#>  $ xmp:CreatorTool                            : chr "TeX"
#>  $ xmpTPg:NPages                              : chr "37"
```

Similar Packages
----------------

There is some overlap with many other text parsers, such as the R interface to antiword (See: <https://github.com/ropensci/antiword>). Listing all of them would take a huge amount of space, since Apache Tika processes over a thousand file types (See: <https://tika.apache.org/>). The main difference is that instead of specializing on a single format, Tika integrates dozens of specialist libraries from the Apache Foundation. Tika's unified approach offers a bit less control, and in return eases the parsing of digital archives filled with possibly unpredictable file types.

In September 2017, github.com user *kyusque* released `tikaR`, which uses the `rJava` package to interact with Tika (See: <https://github.com/kyusque/tikaR>). As of writing, it provided similar text and metadata extraction, but only `xml` output.

Back in March 2012, I started a similar project to interface with Apache Tika. My code also used low-level functions from the `rJava` package. I halted development after discovering that the Tika command line interface (CLI) was easier to use. My empty repository is at <https://r-forge.r-project.org/projects/r-tika/>.

I chose to finally develop this package after getting excited by Tika's new 'batch processor' module, written in Java. The batch processor has very good efficiency when processing tens of thousands of documents. Further, it is not too slow for a single document either, and handles errors gracefully. Connecting `R` to the Tika batch processor turned out to be relatively simple, because the `R` code is simple. It uses the CLI to point Tika to the files. Simplicity, along with continuous testing, should ease integration. I anticipate that some researchers will need plain text output, while others will want `json` output. Some will want multiple processing threads to speed things up. These features are now implemented in `rtika`, although apparently not in `tikaR` yet.

Code of Conduct
---------------

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS
0.1.0 (2018-01-19)
=========================
-Initial release.
-R interface to Apache Tika batch processing CLI, found to be the most efficient CLI option.
-tika function returns processing results as a character vector.
-includes the Tika App .jar. Tika source is available at: https://github.com/apache/tika

0.1.1 (2018-01-23)
=========================
-allows the user to input the URLs and file paths of documents. URLs will be downloaded first to a temporary directory. The previous interface has been changed.
-the Tika License file is included in the source.
-added a vignette on basic text processing using the library.

0.1.2 (2018-01-30)
=========================
-added many tests to increase code coverage dramatically.
-removed the n_chars parameter in favor of using  file.size() internally.
-integrated the covr package.
-for Windows users, the curl package is recommended to prevent base R download.file from corrupting files.
-removed onLoad.R that checked java version in order to speed package loading and simplify code coverage testing.

0.1.3 (2018-02-04)
=========================
-added convenience functions that advertise output format: tika_xml(), tika_json(), tika_text(), tika_html().
-README examples use magrittr pipe.

0.1.4 (2018-02-15)
=========================
-Because it is too big for CRAN, removed the Tika .jar file.
-Added the Tika .jar to a new tikajar package on github.
-Put the ropensci review badge on the tikajar package also, since its an essential component of this package.
-Updated DESCRIPTION, documentation and .travis.yml to reflect the new installation routine.

0.1.5 (2018-02-15)
=========================
-Added dependency on 'sys' package because the 'system2' function was causing intermittent errors by ending tika in mid process.
-Added startup check of the java version, using .onLoad() call to 'java -version'
-Removed redundant conversion to UTF-8, because the Tika batch routine is already outputting UTF-8. 
-Increased the speed of building packages (fewer downloads needed for testing, and the examples do not run).
-Added Code of Conduct to CONDUCT.md file
-Set default 'cleanup' attribute to TRUE.

R/tika.R
#' R Interface to 'Apache Tika'
#'
#' Extract text or metadata from over a thousand file types.
#' Get either plain text or structured \code{XHTML}.
#' Metadata includes \code{Content-Type}, character encoding, and Exif data from 
#' jpeg or tiff images. See the long list of supported file types: 
#' \url{https://tika.apache.org/1.17/formats.html}.
#'
#' @param input Character vector describing the paths to the input documents. 
#' Strings starting with 'http://','https://', or 'ftp://' are downloaded to a 
#' temporary directory first. Each file will be read, but not modified.
#' @param output Optional character vector of the output format. The default, 
#' \code{"text"}, gets plain text without metadata. \code{"xml"} and 
#' \code{"html"} get \code{XHTML} text with metadata. \code{"jsonRecursive"} 
#' gets \code{XHTML} text and \code{json} metadata. 
#' \code{c("jsonRecursive","text")} or \code{c("J","t")} get plain text and 
#' \code{json} metadata. See the 'Output Details' section.
#' @param output_dir Optional directory path to save the converted files in. 
#' Tika may overwrite files so an empty directory is best. See the 'Output 
#' Details' section before using.
#' @param java Optional command to invoke Java. For example, it can be the full 
#' path to a particular Java version. See the Configuration section below.
#' @param jar Optional alternative path to a \code{tika-app-X.XX.jar}. Useful 
#' if this package becomes out of date.
#' @param threads Integer of the number of file consumer threads Tika uses. 
#' Defaults to 2.
#' @param args Optional character vector of additional arguments passed to Tika, 
#' that may not yet be implemented in this R interface, in the pattern of 
#' \code{c('-arg1','setting1','-arg2','setting2')}. Available arguments include 
#' \code{-timeoutThresholdMillis} (Number of milliseconds allowed to a parse 
#' before the process is killed and restarted), \code{-maxRestarts} (Maximum 
#' number of times the watchdog process will restart the child process), 
#' \code{-includeFilePat} (Regular expression to determine which files to 
#' process, e.g. \code{"(?i)\.pdf"}), \code{-excludeFilePat}, and 
#' \code{-maxFileSizeBytes}. These are documented in the .jar --help command.
#' @param quiet Logical if Tika command line messages and errors are to be 
#' suppressed. Defaults to TRUE.
#' @param cleanup Logical to clean up temporary files after running the command, 
#' which can accumulate. Defaults to TRUE. They are in \code{tempdir()}. These 
#' files are automatically removed at the end of the R session even if set to 
#' FALSE.
#' @param lib.loc Optional character vector describing the library paths 
#' containing \code{curl} and \code{data.table} packages. Normally, it's best to 
#' install these and leave this parameter alone. The parameter is included 
#' mainly for package testing.
#' @return A character vector in the same order and with the same length as 
#' \code{input}. Unprocessed files are \code{as.character(NA)}. See the Output 
#' Details section below.
#' @examples
#' \donttest{
#' #' #extract text
#' input= 'https://cran.r-project.org/doc/manuals/r-release/R-data.pdf'
#' text = tika(input)
#' cat(substr(text[1],45,450))
#'
#' #get metadata
#' if(requireNamespace('jsonlite')){
#'   json = tika(input,'J') # capital J is shortcut for jsonRecursive
#'
#'   metadata = jsonlite::fromJSON(json[1])
#'   str(metadata) #meta meta-data
#'
#'   metadata$'Content-Type' # [1] "application/pdf"
#'   metadata$producer # [1] "pdfTeX-1.40.18"
#'   metadata$'Creation-Date' # [1] "2017-11-30T13:39:02Z"
#' }
#' }
#' @section Output Details:
#' If an input file did not exist, could not be downloaded, was a directory, or 
#' Tika could not process it, the result will be \code{as.character(NA)} for 
#' that file.
#'
#' By default, \code{output = "text"} and this produces plain text with no 
#' metadata. Some formatting is preserved in this case using tabs, newlines and 
#' spaces.
#'
#' Setting \code{output} to either \code{"xml"} or the shortcut \code{"x"} will 
#' produce a strict form of \code{HTML} known as \code{XHTML}, with metadata in 
#' the \code{head} node and formatted text in the \code{body}.
#' Content retains more formatting with \code{"xml"}. For example, a Word or 
#' Excel table will become a HTML \code{table}, with table data as text in 
#' \code{td} elements. The \code{"html"} option and its shortcut \code{"h"} 
#' seem to produce the same result as \code{"xml"}.
#' Parse XHTML output with \code{xml2::read_html}.
#'
#' Setting \code{output} to \code{"jsonRecursive"} or its shortcut \code{"J"} 
#' produces a tree structure in `json`. Metadata fields are at the top level. 
#' The \code{XHTML} or plain text will be found in the \code{X-TIKA:content} 
#' field. By default the text is \code{XHTML}. This can be changed to plain 
#' text like this: \code{output=c("jsonRecursive","text")} or 
#' \code{output=c("J","t")}. This syntax is meant to mirror Tika's. Parse 
#' \code{json} with \code{jsonlite::fromJSON}.
#'
#'  If \code{output_dir} is specified, then the converted files will also be 
#'  saved to this directory. It's best to use an empty directory because Tika 
#'  may overwrite existing files. Tika seems to add an extra file extension to 
#'  each file to reduce the chance, but it's still best to use an empty 
#'  directory. The file locations within the \code{output_dir} maintain the same 
#'  general path structure as the input files. Downloaded files have a path 
#'  similar to the `tempdir()` that R uses. The original paths are now relative 
#'  to \code{output_dir}.  Files are appended with \code{.txt} for the default 
#'  plain text, but can be \code{.json}, \code{.xml}, or \code{.html} depending 
#'  on the \code{output} setting. One way to get a list of the processed files 
#'  is to use \code{list.files} with \code{recursive=TRUE}.
#'  If \code{output_dir} is not specified, files are saved to a volatile temp 
#'  directory named by \code{tmpdir()} and will be deleted when R shuts down. 
#'  If this function will be run on very large batches repeatedly, these 
#'  temporary files can be cleaned up every time by adding
#'  \code{cleanup=TRUE}.
#' @section Background:
#' Tika is a foundational library for several Apache projects such as the Apache 
#' Solr search engine. It has been in development since at least 2007. The most 
#' efficient way I've found to process many thousands of documents is Tika's 
#' 'batch' mode, which is the only mode used in `rtika`. There are potentially
#' more things that can be done, given enough time and attention, because 
#' Apache Tika includes many libraries and methods in its .jar file found in 
#' the \strong{tikajar} dependency. The source is available at: 
#' \url{https://tika.apache.org/}.
#' @section Configuration:
#' This package depends on the \code{sys} and \code{tikajar} packages. The later 
#' contains the \code{tika-app-X.XX.jar}. This jar works with Java 7. Tika in 
#' mid-2018 needs Java 8, so it's best to install Java 8.
#'
#' By default, this R package internally invokes Java by calling the \code{java} 
#' command from the command line. To specify the path to a particular Java 
#' version, set the path in the \code{java} attribute of the \code{tika} 
#' function.
#'
#' Other command line arguments can be set with \code{args}. See the options 
#' for version 1.17 here: \url{https://tika.apache.org/1.17/gettingstarted.html}
#'
#' Having the \code{data.table} package installed will slightly speed up the 
#' communication between R and Tika, but especially if there are hundreds of 
#' thousands of documents to process.
#'
#' The \code{curl} package downloads files quickly, if the user includes urls 
#' in the \code{input}. In testing, \code{curl} is required on Windows to avoid 
#' errors, and more work may still be needed to make Windows parse reliably.
#' @export
tika <- function(input,
                 output = c("text", "jsonRecursive", "xml", "html")[1],
                 output_dir = "",
                 java = "java",
                 jar = tikajar::tikajar(),
                 threads = 2,
                 args = character(),
                 quiet = TRUE,
                 cleanup = TRUE,
                 lib.loc = .libPaths()) {
    
    # Special thanks to Hadley the git tutorial at:
    # http://r-pkgs.had.co.nz/git.html
    # Useful functions: 
    # devtools::test(); 
    # devtools::build_vignettes() ; 
    # system('R CMD Rd2pdf ~/rtika')
    
    # TODO:  memory setting with java -Xmx1024m -jar. 
    # Probably also adjust child process -JXmx4g
    
    
    # Parameter sanity check --------------------------------------------
    stopifnot(
        class(input) == "character",
        length(input) > 0,
        class(output) == "character",
        length(output) > 0,
        length(output) <= 2,
        class(output_dir) == "character",
        length(output_dir) == 1,
        class(java) == "character",
        length(java) == 1,
        class(jar) == "character",
        length(jar) == 1,
        class(threads) %in% c("integer", "numeric"),
        class(args) == "character",
        class(quiet) == "logical",
        class(cleanup) == "logical",
        class(lib.loc) == "character"
    )
    # TODO: consider a config file for each request with fine grained control over parsers.
    # see: https://tika.apache.org/1.17/configuring.html
    
    # Define return variable structure  -----------------------------------------
    # output will be character vector the same length as input, with initial NAs ...
    out <- character(length(input))
    failure <- as.character(NA)
    out[out == ""] <- failure
    
    # Parameter tidying  --------------------------------------------------------
    
    # output_flag is output format for Tika command line
    output_flag <- character()
    output_flag <- ifelse(any(output %in% c("jsonRecursive", "J", "-J")),
                          "-J",
                          output_flag) # goes first
    output_flag <- c(output_flag, ifelse(any(output %in% c("text", "t", "-t")),
                                         "-t", 
                                         NA))
    output_flag <- c(output_flag, ifelse(any(output %in% c("xml", "x", "-x")),
                                         "-x",
                                         NA))
    output_flag <- c(output_flag, ifelse(any(output %in% c("html", "h", "-h")),
                                         "-h",
                                         NA))
    output_flag <- as.character(stats::na.omit(output_flag))
    
    
    # output_dir parameter stores tika's processed files. 
    # If it doesn't exist, create one in the temp directory
    if (output_dir == "") {
        # The filenames are guaranteed not to be currently in use.
        output_dir <- tempfile("rtika_dir") 
        dir.create(output_dir)
    } else {
        # if an output directory is provided, check it exists.
        output_dir <- normalizePath(output_dir, mustWork = TRUE, winslash = "/")
        # Must be very careful writing to any directory outside of temp directory 
        # Idea here it to check its not the root directory
        if (output_dir == normalizePath("/", winslash = "/")) {
            stop("Output directory should not be the same as the system root.")
        }
    }
    
    # input parameter may contain URLs. Download if needed
    toDownload <- grep("^(http[s]?:/|ftp:/|file:/)", input, ignore.case = TRUE)
    if (length(toDownload) > 0) {
        if (requireNamespace("curl", quietly = TRUE, lib.loc = lib.loc)) {
            .rtika_fetch <- function(url) {
                ret <- tempfile("rtika_file")
                req <- tryCatch(curl::curl_fetch_disk(url, ret), 
                                error = function(e) return(list(status_code = 418)))
                if (req$status_code != 200) {
                    warning("Could not download with curl::curl_fetch_disk: ", url)
                    return(as.character(NA))
                }
                # TODO: consider adding file affix recording document type (e.g. .docx) 
                # With OS X and ubuntu, content-type is recorded nicely in extended attributes.
                # Check with the cmd: file --mime-type -b rtika_file*
                # However, on Windows, content type is not recorded apparently. 
                # Neither download.file and curl help
                # Curl *does* put content-type in the header provided by the download server
                # headers = parse_headers(req$headers)
                # ctype <- headers[grepl("^content-type", headers, ignore.case = TRUE)]
                # These can be mapped to a database of file affixes using these:
                # * https://raw.githubusercontent.com/apache/tika/master/tika-core/src/main/resources/org/apache/tika/mime/tika-mimetypes.xml
                # * ubuntu has a list in: /etc/mime.types
                # I assume content-type headers are much more reliable than the URL endpoint.
                # This could be made into a 'robust_download' function, that also does retries
                
                return(ret)
            }
        } else {
            if (.Platform$OS.type == "windows") 
            { warning("Please install the 'curl' package.") }
            .rtika_fetch <- function(url) {
                ret <- tempfile("rtika_file")
                req <- tryCatch(utils::download.file(url, ret, method = "libcurl"), 
                                error = function(e) {
                                    return(1)
                                })
                if (req != 0) {
                    warning("Could not download with utils::download.file: ", url)
                    return(as.character(NA))
                }
                return(ret)
            }
        }
        # input parameter adds downloaded file paths (if not downloaded, rtika_download produces NAs)
        urls <- input[toDownload]
        tempfiles <- unlist(lapply(urls, .rtika_fetch))
        input[toDownload] <- tempfiles
    }
    
    # Input can have issues. 
    # Check if file exist, are not directories, are not NA, and were downloaded
    file_exists <- !is.na(input) & file.exists(input) & !dir.exists(input)
    if (!any(file_exists)) {
        warning("No files could be found.")
        return(out)
    }
    
    # inputFiles and fileList will contain files for Tika to process  --------------------------
    
    inputFiles <- normalizePath(input[file_exists], winslash = "/")
    # Tika expects files to be relative to root
    # NB: These paths will be READ, NOT written to!
    root <- normalizePath("/", winslash = "/")
    inputFiles <- sub(root, "", inputFiles, fixed = TRUE)
    
    # fileList is a delimited versiob if inputFiles that will be passed to Tika.
    # File paths containing both commas and quote characters appear to work with the settings below.
    fileList <- tempfile("rtika_file")
    if (requireNamespace("data.table", quietly = TRUE, lib.loc = lib.loc)) {
        data.table::fwrite(data.table::data.table(inputFiles),
                           fileList, row.names = FALSE,
                           col.names = FALSE,
                           sep = ",",
                           quote = FALSE)
    } else {
        utils::write.table(inputFiles,
                           fileList, row.names = FALSE,
                           col.names = FALSE,
                           sep = ",",
                           quote = FALSE)
    }
    # After the file is created, make sure it exists
    fileList <- normalizePath(fileList, mustWork = TRUE, winslash = "/")
    
    # java call  --------------------------------------------------------
    
    
    if (.Platform$OS.type == "windows") {
        # Windows java requires quoting for paths, but OS X and Ubuntu java run into problems with shQuote. 
        java_args <- c("-Djava.awt.headless=true", "-jar", shQuote(jar), 
                       "-numConsumers", as.integer(threads), args, output_flag, 
                       "-i", shQuote(root), "-o", shQuote(output_dir), 
                       "-fileList", shQuote(fileList))
    } else {
        java_args <- c("-Djava.awt.headless=true", "-jar", jar, "-numConsumers", 
                       as.integer(threads), args, output_flag, "-i", root, "-o", 
                       output_dir, "-fileList", fileList)
    }
    
    # Compared to system2, sys is somehow much quicker when making the call to java.
    # TODO: catch java errors better. 
    
    sys::exec_wait(cmd = java[1], args = java_args, std_out = !quiet, 
                   std_err = !quiet)
    
    # retrieve results  --------------------------------------------------------
    
    output_file_affix <- character()
    output_file_affix <- ifelse(any(output %in% c("text", "t", "-t")),
                                ".txt",
                                output_file_affix)
    output_file_affix <- ifelse(any(output %in% c("xml", "x", "-x")),
                                ".xml",
                                output_file_affix)
    output_file_affix <- ifelse(any(output %in% c("html", "h", "-h")),
                                ".html",
                                output_file_affix)
    output_file_affix <- ifelse(any(output %in% c("jsonRecursive", "J", "-J")),
                                ".json",
                                output_file_affix)
    
    # Vectorized readChar
    .rtika_readFile <- function(path) {
        bytes <- file.size(path)
        ifelse(!is.na(bytes), mapply(readChar,
                                     con = path,
                                     nchars = bytes,
                                     useBytes = TRUE),
               as.character(NA))
    }
    
    # Clean & check with normalizePath, with warnings if not processed.
    output_files <- normalizePath(file.path(output_dir,
                                            paste0(inputFiles, 
                                                   output_file_affix)),
                                  winslash = "/")
    
    out[file_exists] <- .rtika_readFile(output_files)
    
    
    # From studying the source code, the Tika batch processor defaults to UTF-8
    # every batch config xml file uses a FSOutputStreamFactory set to UTF-8
    
    # cleanup temp files  -----------------------------------------------------
    if (cleanup) {
        trash_file <- file.path(tempdir(), list.files(tempdir(), 
                                                      pattern = "^rtika_file"))
        if (length(trash_file) > 0) {
            tmp <- file.remove(trash_file)
        }
        
        trash_dir <- file.path(tempdir(), list.files(tempdir(), 
                                                     pattern = "^rtika_dir"))
        if (length(trash_dir) > 0) { 
            # if(.Platform$OS.type=='windows'){
            # On windows, deleting dir does not seem to work unless force=TRUE.
            # but do not want to force this.
            #   tmp=  unlink( trash_dir, recursive =TRUE, force=TRUE) 
            # } else {
            tmp <- unlink(trash_dir, recursive = TRUE)
            # }
        }
    }
    
    return(out)
}
